### ISSUE TITLE
Reaction run()

### ISSUE THREAD

**Issue by mindplay-dk** [zheksoon/dipole#7]
As covered in the documentation, you have to manually `.run()` a reaction the first time:

```js
const r = reaction(() => {  
    console.log(`Sum of ${a.get()} and ${b.get()} is ${sum.get()}`)  
})  
r.run()     // first run should be triggered manually  
```

The documentation doesn't really explain why that is?

The thing that puzzles me is, what do you actually *have*, if you haven't called the `run` method yet?

You don't really have a reaction yet - if it hasn't run the function, it doesn't know the function's dependencies, and so it can't actually react to anything yet.

When is an "inactive" reaction useful? Or why isn't the reaction automatically run the first time?

**Comment by zheksoon** [comment]
> The thing that puzzles me is, what do you actually have, if you haven't called the run method yet?

So this is simple - until you run reaction all you have is Reaction object with fields set to point to the reaction body function. After `run()` reaction subscribes to observable/computed values, so `_subscriptions` field is filled with its dependencies after that.

> When is an "inactive" reaction useful? Or why isn't the reaction automatically run the first time?

Just check out my recent commit to react bindings (still not done yet, was quite short on time):
https://github.com/zheksoon/dipole-react/blob/9a431a3bbe7d13af346b464b453ec77bc3591393/index.ts#L11

`useMemo` creates an empty reaction instance, which is ran later at component return. Note that Reaction's `run()` method is not parameter-less - it passes whatever you want to reaction body function and also returns its execution result. In this case the params are component props (and obsolete `context`), and return result is JSX markup of component. 

The difference from the documentation use case is Reaction's `manager` param that was designed exactly for this kind of use cases - when reaction's dependencies change, the `manager` function runs instead of reaction body, and its task is to call `reaction.run()` somehow. Following the code, the manager function is `() => triggerUpdate({})`, which makes React re-execute the component's body and run the `r.run()` at the end. 

Similary, `reactionClass` wraps class component and maybe even simpler to understand.

**Comment by mindplay-dk** [comment]
But it seems the only point of this is to defer the initialization of the reaction?

The constructor itself doesn't *do* anything, other than binding variables to an object - so you essentially just have a dysfunctional instance, until you trigger the actual "constructor", which is the `run` method.

What you get with this pattern at construction time is just variables bound to an object - which seems no different from just binding variables to a closure.

Generally, if you want to defer the initialization of something, just don't create it yet - for example:

```js
const defer = fn => {
  let resolved;

  return () => resolved || (resolved = fn());
}

function notYet() {
  console.log("BUT NOW!"); // runs only once

  return "HELLO";
}

const later = defer(notYet); // doesn't run

console.log("not yet...");

console.log(later()); // runs the first time

console.log(later()); // but runs only once
```

Output:

```
not yet...
BUT NOW!
HELLO
HELLO
```

Here, I'm deferring the creation of a string `HELLO`, but this pattern should work for anything?

**Comment by zheksoon** [comment]
Your example with deferred initialization makes sense for some caching computation, but in case of Reaction class there is no cache - calling `run()` method should always execute reaction body, resubscribe to its dependencies and return execution result. 

Another reason for the `reaction` factory function and `Reaction` constructor not doing anything but fields initialization is the fact that all core classes of dipole follow the same design decision - be as simple and predictable as possible, with no hidden implicit behavior.

**Comment by mindplay-dk** [comment]
I think you misunderstood, I'm not suggesting you use this pattern in the library. 🙂

What I'm suggesting is, a *consumer* can use this pattern - for example, in your React adapter, I don't think you need to actually construct the reaction before you're ready to initialize it. You can use this pattern to defer the creation until you're ready.

In other words, this library (or any library) doesn't need to be concerned with timing issues on the consumer's side - the need to defer the initialization of a reaction is an implementation detail of that library that it can solve by, simply, not constructing the reaction until it's ready to.

A reaction doesn't have any function at all until you call `run` - it's a dead object with no reason to exist yet.

Having an constructor that generates a *working instance* definitely seems simpler and more predictable to me.

**Comment by mrjjwright** [comment]
I was able to integrate dipole with https://github.com/Freak613/1more using this mechanism.  I am hoping dipole will be the most trustworthy lib I have ever used in the reactivity space as  I have chosen it for it's transparency, so far it's working great and I love the Mobx like primitives with which I am familiar!

**Comment by mrjjwright** [comment]
One question: when nesting Reactions because of nested components, either in React or any kind of component tree, how does dipole make sure that all the components up higher in the tree are not updated when lower level reactions fire, since lower level reactions would be called from the higher level reactions run methods?

**Comment by tsugabloom** [comment]
I like the idea of manually starting a reaction; it feels like there's choice and control. I also do this in my reactive library Wire because I want creation and execution to be two separate stages so I can give the most flexibility to people who need to modify the reaction such as their properties - I do these modifications on shutdown/offline reactions before they're ever started do avoid wasted work.

Consumers who are very eager to have a single function that does both a create+start can easily write a wrapper function. Meanwhile it's harder to write a proper wrapper that can _split_ an existing create+start. For instance, your "defer" approach above won't work for people already doing `if (x instanceof Reaction)` checks.
 
I've heard a similar "it's useless" argument for DOM nodes when people say it's useless that you can create a DOM node that's not yet mounted to the DOM. Sure... but I think it's _useful to someone_, while removing/merging that split functionality will likely _impede someone_ trying to have a more fine-grained control over the code.

**Comment by mindplay-dk** [comment]
> I like the idea of manually starting a reaction; it feels like there's choice and control. I also do this in my reactive library Wire because I want creation and execution to be two separate stages so I can give the most flexibility to people who need to modify the reaction such as their properties

I haven't seen that library.

But in Dipole, the `Reaction` class doesn't have any public properties.

So this feeling of choice and control, are you sure it's more than just perception? 🙂

*EDIT*: I guess one argument I could think of, is if somebody wants to *extend* the `Reaction` class with mutable properties for some reason...

*EDIT, EDIT*: ... but in that case, they would need to replace the `reaction` function as well... so perhaps it would be reasonable to initialize the `Reaction` in the `reaction` function, given that the base `Reaction` class has no mutable properties, and you'd have to write your own factory-function anyhow? 🤔
